core/server/master/src/main/java/alluxio/master/file/meta/InodeTreePersistentState.java

```diff
@@ -677,35 +685,195 @@ public void restoreFromCheckpoint(CheckpointInputStream input) throws IOExceptio
+              Iterable<? extends Inode> children = mInodeStore.getChildren(mDirInode.asDirectory());
+              children.forEach((child) -> {
+                try {
+                  if (child.isDirectory()) {
+                    // Insert directory for further branching.
+                    mDirectoriesToIterate.put(child);
+                  } else {
+                    // Buffer current file as JournalEntry
+                    mEntryBuffer.put(child.toJournalEntry());
+                  }
+                } catch (InterruptedException ie) {
+                  // Continue interrupt chain.
+                  Thread.currentThread().interrupt();
+                  throw new RuntimeException("Thread interrupted while enumerating a dir.");
+                }
+              });
+            } catch (InterruptedException ie) {
+              // Continue interrupt chain.
+              Thread.currentThread().interrupt();
+              throw new RuntimeException("Thread interrupted while enumerating on a dir.");
+            } finally {
+              try {
+                // Signal completion
+                mCompletionQueue.put(true);
+              } catch (InterruptedException ie) {
+                // Continue interrupt chain.
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(
+                    "Thread interrupted while signaling completion for an enumeration.");
+              }
+            }
+          }
+        }
+
+        // Create coordinator task.
+        mCoordinatorExecutor.submit(() -> {
+          try {
+            AtomicInteger runningCrawlerCount = new AtomicInteger(0);
+            // Loop as long as there is a dir to branch or there are active enumeration thread.
+            while (runningCrawlerCount.get() > 0 || !mDirectoriesToIterate.isEmpty()) {
+
+              if (!mDirectoriesToIterate.isEmpty()) {
+                // There is a dir to enumerate.
+                runningCrawlerCount.incrementAndGet();
+                mCrawlerExecutor.submit(new DirectoryCrawler(mDirectoriesToIterate.take()));
+              } else {
+                // No dirs but there are active threads.
+                mCompletionQueue.take();
+                runningCrawlerCount.decrementAndGet();
+              }
+            }
+          } catch (InterruptedException ie) {
+            // Continue interrupt chain.
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("Thread interrupted while waiting for enumeration threads.");
+          } finally {
+            // Signal completion of buffering.
+            mBufferingCompleted.set(null);
+            // Crawling completed.
+            mCrawlerExecutor.shutdown();
+          }
+        });
+        // Will be shutdown after current coordinator task is complete.
+        mCoordinatorExecutor.shutdown();
+      }
+
       @Override
       public boolean hasNext() {
```

Comment:
Can we make this blocking so that `hasNext()` is true iff there really is a next element? It will make the iterator much more pleasant to work with